seed/data_importer/match.py

```diff
@@ -728,32 +746,40 @@ def states_to_views(unmatched_state_ids, org, access_level_instance, cycle, Stat
     )
 
 
-def link_states(states, ViewClass, cycle, highest_ali, sub_progress_key):  # noqa: N803
-    """
-    Run each of the given -States through a linking round.
-
-    For details on the actual linking logic, please refer to the
-    match_merge_link() method.
-    """
+def link_states(states, ViewClass, cycle, highest_ali, sub_progress_key, tuple_values, view_lookup, match_lookup):  # noqa: N803
+    class_name = "property" if ViewClass == PropertyView else "taxlot"
 
+    # set up the progress bar
     sub_progress_data = update_sub_progress_total(100, sub_progress_key)
+    batch_size = math.ceil(len(states) / 100)
 
     linked_views = []
     unlinked_views = []
     invalid_link_states = []
     unlinked_states = []
-
-    batch_size = math.ceil(len(states) / 100)
     for idx, state in enumerate(states):
+        # get matches
+        existing_state_matches = match_lookup.get(tuple(getattr(state, c) for c in tuple_values), [])
+        existing_views_matches = [view_lookup[x["id"]] for x in existing_state_matches]
+
+        # ensure ali is correct
+        view = ViewClass.objects.filter(state_id=state.id, cycle_id=cycle.id).select_related(f"{class_name}__access_level_instance").first()
```

Comment:
Could this query could be sped up by using states realtion to view
`state.propertyview_set.filter(cycle_id=cycle.id).select_related(f"{class_name}__access_level_instance").first()`

side bar: why is PropertyState a 1-M with PropertyView? It seems a 1-1 relationship would be more appropriate ([see line 407 in models/properties.py](https://github.com/SEED-platform/seed/blob/develop/seed/models/properties.py#L407))

Reply:
I''m unsure if that speeds it up, but it can't hurt!

> side bar: why is PropertyState a 1-M with PropertyView? It seems a 1-1 relationship would be more appropriate (error log if ps has more than 1 pv)

My fake answer is that hypothetically, technically, kinda sorta, two views of a single property could want to be represented by the same state, should there only be one record of info for both 2012 and 2013 or something, idk. You'd have to do that programmatically, the UI (and API I think) would never allow that.

My real answer is I believe the previous owners of this codebase didn't know how to code 1-1s. If you look for it you start to see those oversights around.

Reply:
Yeah, I didn't see any meaningful difference while testing both versions of that query.

Thanks for both answers :), I fear the latter may be true.